spec/function.dd

```diff
@@ -173,124 +173,246 @@ int foo();
+$(UL
+$(LI read and write the floating point exception flags;)
+$(LI read and write the floating point mode flags, as long as those
+flags are restored to their initial state upon function entry;)
+$(LI perform impure operations in statements that are in a
+$(GLINK2 version, ConditionalStatement)
+controlled by a $(GLINK2 version, DebugCondition).)
+)
 
-        $(P As a concession to practicality, a pure function can also:)
+---
+import std.stdio;
+pure int cube(int i)
+{
+    debug writeln("cube() called"); // ok, impure code allowed in debug statement
+    return i * i * i;
+}
+---
 
-        $(UL
-        $(LI read and write the floating point exception flags)
-        $(LI read and write the floating point mode flags, as long as those
-        flags are restored to their initial state upon function entry)
-        $(LI perform impure operations in statements that are in a
-        $(GLINK2 version, ConditionalStatement)
-        controlled by a $(GLINK2 version, DebugCondition).)
-        )
+$(P A pure function can throw exceptions (unless it has the `nothrow` attribute).)
 
-        $(P A pure function can throw exceptions.)
+$(P Note: The rules above do not preclude modification of data obtained through parameters. Unlike
+traditional functional programming languages, D's pure functions allow modification of the caller's
+state through their mutable parameters. This definition of mutable functions is more relaxed than
+the one traditionally employed by pure functional languages because it allows a pure function to
+use state mutation, as long as all state is created internally or reachable through its arguments.
+In particular, a pure function may allocate memory by means of e.g. `new` or `malloc` without these
+being special cases. Example:)
 
 ---
-import std.stdio;
-int x;
-immutable int y;
-const int* pz;
-
-pure int foo(int i,
-             char* p,
-             const char* q,
-             immutable int* s)
-{
-    debug writeln("in foo()"); // ok, impure code allowed in debug statement
-    x = i;   // error, modifying global state
-    i = x;   // error, reading mutable global state
-    i = y;   // ok, reading immutable global state
-    i = *pz; // error, reading const global state
-    return i;
+pure size_t foo(int[] arr) { arr[] += 1; return arr.length; }
+int[] a = [1, 2, 3];
+foo(a);
+assert(a == [2, 3, 4]);
+---
+
+$(P Pure functions fall in exactly one of two categories: $(I weakly pure) and $(I strongly pure).
+There are no dedicated annotations for these; instead, a set of rules applied to the function
+declaration determines its category.)
+
+$(H3 Weak Purity)
+
+$(P A function annotated with `pure` is weakly pure if and only if at least one of the
+following conditions are met:)
+
+$(OL
+$(LI One or more parameter types are incompletely defined. Example:)
+
+---
+struct S; // defined elsewhere
+// These functions are weakly pure:
+pure int fun(S object);
+pure int fun(immutable S object);
```

Comment:
Why is this not strongly pure? Clearly it does not have mutable indirections because of transitivity of immutability.

Reply:
@tgehr this is to leave us room for `__metadata`.

Reply:
It contradicts existing behavior:
```
struct S;
int[] foo(immutable S* s)pure;
ink 
immutable(int[]) bar(immutable S* s){
	return foo(s); // ok
}
```
`__metadata` shouldn't influence strong purity anyway. Otherwise we can't actually implement lazy evaluation.

Reply:
@tgehr I don't understand the example. Can you please clarify?

>__metadata shouldn't influence strong purity anyway. Otherwise we can't actually implement lazy evaluation correctly.

I don't understand this either, please clarify. Thanks!

Reply:
> @tgehr I don't understand the example. Can you please clarify?
> ...

The function `foo` computes an `int[]` from a pointer to an immutable instance of an opaque datatype. The result converts to `immutable` implicitly. The PR says this is not the case.
 
> > __metadata shouldn't influence strong purity anyway. Otherwise we can't actually implement lazy evaluation correctly.
> 
> I don't understand this either, please clarify. Thanks!

Simplest example is you have an `immutable` data type and you change some field from being eagerly initialized to being lazily initialized. Factory functions that take an `immutable` instance of your type will break, because their results suddenly cannot be implicitly converted to `immutable` anymore.

Reply:
@tgehr thanks, I see. The short of it is I don't know how to define `__metadata` and `pure` so as to work together properly without blowing complexity up. If you have any ideas, they'd be welcome.

The longer of it is I'm willing to impose additional restrictions on unknown data types if that makes matters simple, even if we need to break code that works in obscure cases. Optimizing strongly pure functions is low-impact anyway (we have zero such optimizations right now and there's no blood in the streets), and it's very rare that types are incompletely defined in D to start with.